File: c++/71_Simplify_Path.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <stack>

using namespace std;

class Solution {
public:
    string simplifyPath(string path) {
    	string res, part;
    	vector<string> vec;
    	path.push_back('/');
    	for (int i = 0; i < path.size(); i++) {
    		if (path[i] != '/')
    			part.push_back(path[i]);
    		else if (path[i] == '/') {  			
    			if (part.empty())
    				continue;
    			else if (part == "..") {
    				if (vec.empty()) {
    					part.clear();
    					continue;
    				}
    				vec.pop_back();
    			}
    			else if (part != ".")
    				vec.push_back(part);
    			part.clear();
    		}
    	}

		for (int i = 0; i < vec.size(); i++) {
    		res += "/" + vec[i];
    	}
    	if (res.empty())
    		res = "/";
    	return res;
    }
};


int main() {
	Solution s;
	cout << s.simplifyPath("/home/../../..") << endl;
}

File: python/71_Simplify_Path.py
#!/usr/bin/python

class Solution(object):
    def simplifyPath(self, path):
    	path += "/"
    	res, part = "", ""
    	vec = []
    	for i in range(len(path)):
    		if path[i] != '/':
    			part += path[i]
    		else:
    			if part == "":
    				continue
    			elif part == "..":
    				if vec == []:
    					part = ""
    					continue
    				vec.pop()
    			elif part != ".":
    				vec.append(part)
    			part = ""
    	for i in range(len(vec)):
    		res += "/" + vec[i]
    	if vec == []:
    		res = "/"
    	return res



def main():
    s = Solution()
    print s.simplifyPath("/...")

if __name__ == "__main__":
    main()
